syft/pkg/cataloger/golang/parse_go_bin.go

```diff
@@ -39,57 +63,112 @@ func newGoBinaryPackage(name, version, h1Digest, goVersion, architecture string,
 	return p
 }
 
-func parseGoBin(location source.Location, reader io.ReadCloser, opener exeOpener) (pkgs []pkg.Package, err error) {
-	var exes []exe
-	// it has been found that there are stdlib paths within openExe that can panic. We want to prevent this behavior
-	// bubbling up and halting execution. For this reason we try to recover from any panic and return an error.
-	defer func() {
-		if r := recover(); r != nil {
-			err = fmt.Errorf("recovered from panic while parse go binary at %q: %+v", location.RealPath, r)
+func setArch(readers []io.ReaderAt, builds []*debug.BuildInfo) {
+	if len(readers) != len(builds) {
+		log.Errorf("golang cataloger: bin parsing: number of builds and readers doesn't match")
+		return
+	}
+
+	for _, build := range builds {
+		if getGOARCH(build.Settings) != "" {
+			return
+		}
+		break
+	}
+
+	for i, r := range readers {
+		a, err := getGOARCHFromBin(r)
+		if err != nil || a == "" {
+			log.Warnf("golang cataloger: bin parsing: getting arch from binary: %v", err)
+			continue
 		}
-	}()
 
-	// Identify if bin was compiled by go
-	exes, err = opener(reader)
-	if err != nil {
-		return pkgs, err
+		builds[i].Settings = append(builds[i].Settings, debug.BuildSetting{Key: GOARCH, Value: a})
 	}
+}
 
-	for _, x := range exes {
-		goVersion, mod := findVers(x)
-		pkgs = append(pkgs, buildGoPkgInfo(location, mod, goVersion, x.ArchName())...)
+func getGOARCH(settings []debug.BuildSetting) string {
+	for _, s := range settings {
+		if s.Key == GOARCH {
+			return s.Value
+		}
 	}
-	return pkgs, err
+
+	return ""
 }
 
-func buildGoPkgInfo(location source.Location, mod, goVersion, arch string) []pkg.Package {
-	pkgsSlice := make([]pkg.Package, 0)
-	scanner := bufio.NewScanner(strings.NewReader(mod))
+func getGOARCHFromBin(r io.ReaderAt) (string, error) {
```

Comment:
this function is a rewrite from a set of functions that were extracted from the standard lib in a previous version of go. Now that this has been rewritten we should try to add more tests for this functionality.

Reply:
added test, and support to scoff files